packages/flutter/lib/src/material/floating_action_button_location.dart

```diff
@@ -213,6 +261,53 @@ class _CenterDockedFloatingActionButtonLocation extends _DockedFloatingActionBut
     final double fabX = (scaffoldGeometry.scaffoldSize.width - scaffoldGeometry.floatingActionButtonSize.width) / 2.0;
     return Offset(fabX, getDockedY(scaffoldGeometry));
   }
+
+  @override
+  String toString() => 'FloatingActionButtonLocation.centerDocked';
+}
+
+double _straddleAppBar(ScaffoldPrelayoutGeometry scaffoldGeometry) {
+  final double fabHalfHeight = scaffoldGeometry.floatingActionButtonSize.height / 2.0;
+  return scaffoldGeometry.contentTop - fabHalfHeight;
+}
+
+class _StartTopFloatingActionButtonLocation extends FloatingActionButtonLocation {
+  const _StartTopFloatingActionButtonLocation();
+
+  @override
+  Offset getOffset(ScaffoldPrelayoutGeometry scaffoldGeometry) {
+    return Offset(_startOffset(scaffoldGeometry), _straddleAppBar(scaffoldGeometry));
+  }
+
+  @override
+  String toString() => 'FloatingActionButtonLocation.startTop';
+}
+
+class _MiniStartTopFloatingActionButtonLocation extends FloatingActionButtonLocation {
+  const _MiniStartTopFloatingActionButtonLocation();
+
+  @override
+  Offset getOffset(ScaffoldPrelayoutGeometry scaffoldGeometry) {
+    // We have to offset the FAB by four pixels because the FAB itself _adds_
+    // four pixels in every direction in order to have a hit target area of 48
+    // pixels in each dimension, despite being a circle of radius 40.
+    return Offset(_startOffset(scaffoldGeometry, offset: 4.0), _straddleAppBar(scaffoldGeometry));
```

Comment:
nit: making the magic 4.0 offset a constant could make the code a bit more self documenting and would avoid having the offsets get out of sync.